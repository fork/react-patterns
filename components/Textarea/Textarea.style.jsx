import styled, { css } from 'styled-components';
import { color, space } from '../../stylesheets';

const StyledTextarea = styled.div`
  label {
    display: block;
    font-weight: bold;
    color: ${color('black')};
    margin-bottom: ${space('m')};
  }

  textarea {
    width: 100%;
    outline: none;
    background: transparent;
    color: ${color('black')};
    font-size: inherit;
    font-weight: normal;
    resize: none;
    position: relative;
    border: 1px solid ${color('black')};
    padding: ${space('m')};

    &::placeholder {
      color: ${color('black')};
      font-weight: normal;
    }

    &::-ms-input-placeholder {
      color: ${color('black')};
      font-weight: normal;
    }

    &:-ms-input-placeholder {
      color: ${color('black')};
      font-weight: normal;
    }

    &:focus {
      border-color: ${color('secondary')};
    }
  }

  .textarea__error {
    display: block;
    color: ${color('highlight')};
  }

  ${props =>
    props.error &&
    css`
      label {
        color: ${color('highlight')};
      }

      textarea {
        border-color: ${color('highlight')};
      }
    `}

  ${props =>
    props.success &&
    css`
      label {
        color: ${color('secondary')};
      }

      textarea {
        border-color: ${color('secondary')};
      }
    `}
`;

export default StyledTextarea;
