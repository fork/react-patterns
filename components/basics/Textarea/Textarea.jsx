import PropTypes from 'prop-types';
import React from 'react';
import StyledTextarea from './Textarea.style';

const Textarea = ({
  id,
  label,
  value,
  required,
  placeholder,
  onChange,
  error,
  success,
  disabled,
  autocomplete,
  rows,
  errorMessage,
  ...props
}) => (
  <StyledTextarea disabled={disabled} error={error} success={success} {...props}>
    <label htmlFor={id}>
      <span className="textarea__labelText">{label}</span>
      {required && <span className="textarea__required">*</span>}
    </label>

    <textarea
      id={id}
      name={id}
      value={value}
      placeholder={placeholder}
      onChange={onChange}
      aria-invalid={error}
      aria-required={required}
      required={required}
      disabled={disabled}
      autoComplete={autocomplete}
      rows={rows}
      {...props}
    />

    {error && errorMessage && <span className="textarea__error">{errorMessage}</span>}
  </StyledTextarea>
);

Textarea.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.string,
  required: PropTypes.bool,
  placeholder: PropTypes.string,
  onChange: PropTypes.func,
  error: PropTypes.bool,
  success: PropTypes.bool,
  disabled: PropTypes.bool,
  rows: PropTypes.number,
  autocomplete: PropTypes.string,
  errorMessage: PropTypes.string
};

Textarea.defaultProps = {
  placeholder: '',
  value: '',
  required: false,
  onChange: null,
  error: false,
  success: false,
  disabled: false,
  rows: 4,
  autocomplete: 'off',
  errorMessage: null
};

export default Textarea;
