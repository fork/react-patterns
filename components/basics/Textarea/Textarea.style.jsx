import styled, { css } from 'styled-components';
import { color, space } from '../../../stylesheets';

const StyledTextarea = styled.div`
  label {
    display: block;
    font-weight: bold;
    color: ${color('black')};
    margin-bottom: ${space('m')};
  }

  textarea {
    width: 100%;
    outline: none;
    background: transparent;
    color: ${color('black')};
    font-size: inherit;
    font-weight: normal;
    resize: none;
    position: relative;
    border: 1px solid ${color('black')};
    padding: ${space('m')};

    &::placeholder {
      color: ${color('black')};
      font-weight: normal;
    }

    &::-ms-input-placeholder {
      color: ${color('black')};
      font-weight: normal;
    }

    &:-ms-input-placeholder {
      color: ${color('black')};
      font-weight: normal;
    }

    &:focus {
      border-color: orange;
    }
  }

  .textarea__error {
    display: block;
    color: ${color('primary')};
  }

  ${props =>
    props.error &&
    css`
      label {
        color: ${color('primary')};
      }

      textarea {
        border-color: ${color('primary')};
      }
    `}

  ${props =>
    props.success &&
    css`
      label {
        color: green;
      }

      textarea {
        border-color: green;
      }
    `}
`;

export default StyledTextarea;
