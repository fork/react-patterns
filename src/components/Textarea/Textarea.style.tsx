import styled, { css } from 'styled-components';
import { color, hideVisually, space } from '../../stylesheets';

const StyledTextarea = styled.div<{ error?: boolean; success?: boolean; hideLabel?: boolean }>`
  label {
    display: block;
    font-weight: bold;
    color: ${color('neutral', 100)};
    margin-bottom: ${space('s')};
    ${props => props.hideLabel && hideVisually()}
  }

  textarea {
    width: 100%;
    outline: none;
    background: transparent;
    color: ${color('neutral', 100)};
    font-size: inherit;
    font-weight: normal;
    resize: none;
    position: relative;
    border: 1px solid ${color('neutral', 100)};
    padding: ${space('m')};

    &::placeholder {
      color: ${color('neutral', 100)};
      font-weight: normal;
    }

    &::-ms-input-placeholder {
      color: ${color('neutral', 100)};
      font-weight: normal;
    }

    &:-ms-input-placeholder {
      color: ${color('neutral', 100)};
      font-weight: normal;
    }

    &:focus {
      border-color: ${color('neutral', 30)};
    }
  }

  .textarea__error {
    display: block;
    color: ${color('warning')};
  }

  ${props =>
    props.error &&
    css`
      label {
        color: ${color('warning')};
      }

      textarea {
        border-color: ${color('warning')};
      }
    `}

  ${props =>
    props.success &&
    css`
      label {
        color: ${color('success')};
      }

      textarea {
        border-color: ${color('success')};
      }
    `}
`;

export default StyledTextarea;
