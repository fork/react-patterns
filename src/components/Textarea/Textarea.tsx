import React from 'react';

import Text from '../Text';

import StyledTextarea from './Textarea.style';

export type TextareaProps = {
  label: string;
  id: string;
  required?: boolean;
  onChange?: (ev: React.ChangeEvent<HTMLTextAreaElement>) => void;
  onFocus?: (ev: React.FocusEvent<HTMLTextAreaElement>) => void;
  onBlur?: (ev: React.FocusEvent<HTMLTextAreaElement>) => void;
  error?: boolean;
  errorMessage?: string;
  disabled?: boolean;
  value?: string;
  placeholder?: string;
  success?: boolean;
  hideLabel?: boolean;
  autocomplete?: string;
  rows?: number;
};

const Textarea = ({
  id,
  label,
  value,
  required,
  placeholder,
  onChange,
  onFocus,
  onBlur,
  error,
  success,
  disabled,
  autocomplete,
  rows,
  hideLabel,
  errorMessage
}: TextareaProps) => (
  <StyledTextarea error={error} success={success} hideLabel={hideLabel}>
    <label htmlFor={id}>
      <Text as="span" variant="form-label">
        {label}
        {required && <span className="textarea__required">*</span>}
      </Text>
    </label>
    <textarea
      id={id}
      name={id}
      value={value}
      placeholder={placeholder}
      onChange={onChange}
      onFocus={onFocus}
      onBlur={onBlur}
      aria-invalid={error}
      aria-required={required}
      required={required}
      disabled={disabled}
      autoComplete={autocomplete}
      rows={rows}
    />

    {error && errorMessage && (
      <div className="textarea__error">
        <Text as="span" variant="copy-small">
          {errorMessage}
        </Text>
      </div>
    )}
  </StyledTextarea>
);

export default Textarea;
