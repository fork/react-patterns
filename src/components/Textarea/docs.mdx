import { Story, Canvas } from '@storybook/addon-docs';

# Textarea

Eine Textarea ermöglicht Nutzer:innen die Eingabe von Text, der länger als eine Zeile ist.

<Canvas>
  <Story id="components-input-textarea--default" />
</Canvas>

## Anatomie

<!-- Optional: schematische Darstellung -->

1. **Label** (optional): Beschreibt den Inhalt der Textarea
2. **Placeholder** (optional): Hinweis, wenn Nutzer:innen noch keinen Text eingegeben haben
3. **Character Count** (optional): Stellt die Anzahl der eingegebenen und der maximal erlaubten Zeichen dar.
4. **Error**: Fehlermeldung

<!--
## Platzierung

Hier kann beschrieben werden, wie eine Textarea in verschiedenen Kontexten platziert wird.
-->
