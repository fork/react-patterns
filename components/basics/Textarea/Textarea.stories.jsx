import React, { useState } from 'react';

import Textarea from './Textarea';

export default {
  title: 'Basics/Textarea',
  component: Textarea
};

const props = {
  id: '2',
  name: 'textarea',
  label: 'Your message',
  placeholder: 'Enter your message...'
};

export const WithoutValue = () => <Textarea {...props} />;
export const WithValue = () => <Textarea {...props} value="Johannes" />;
export const Error = () => (
  <Textarea {...props} error errorMessage="Bitte verfasse eine Nachricht." />
);
export const Success = () => <Textarea {...props} success />;
export const withState = () => {
  const [value, setValue] = useState('');
  return <Textarea {...props} value={value} onChange={e => setValue(e.target.value)} />;
};
