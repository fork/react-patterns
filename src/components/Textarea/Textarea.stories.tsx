import React, { useEffect, useState } from 'react';
import { Meta, Story } from '@storybook/react';

import Textarea, { TextareaProps } from './Textarea';
import docs from './docs.mdx';

export default {
  title: 'Components/Textarea',
  component: Textarea,
  parameters: {
    docs: {
      page: docs
    }
  }
} as Meta;

const Template: Story<TextareaProps> = ({ value, ...args }) => {
  const [val, setValue] = useState(value);

  useEffect(() => {
    setValue(value);
  }, [value]);

  return <Textarea value={val} onChange={e => setValue(e.target.value || '')} {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  id: '2',
  name: 'textarea',
  label: 'Your message',
  placeholder: 'Enter your message...',
  required: true,
  error: false,
  success: false,
  disabled: false,
  value: '',
  errorMessage: ''
};

export const Error = Template.bind({});
Error.args = { ...Default.args, error: true, errorMessage: 'Bitte verfasse eine Nachricht.' };

export const Success = Template.bind({});
Success.args = { ...Default.args, success: true };
