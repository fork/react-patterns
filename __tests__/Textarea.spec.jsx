import React from 'react';
import { shallow } from 'enzyme';

import Textarea from '../components/basics/Textarea';

describe('Textarea', () => {
  test('Textarea should render correctly', () => {
    const component = shallow(<Textarea label="Textarea Label" id="1" />);

    expect(component).toMatchSnapshot();
  });

  test('Textarea with hidden label should render correctly', () => {
    const component = shallow(<Textarea hideLabel label="Textarea Label" id="1" />);

    expect(component).toMatchSnapshot();
  });

  test('Textarea with default value should render correctly', () => {
    const component = shallow(<Textarea hideLabel label="Textarea Label" value="true" id="1" />);

    expect(component).toMatchSnapshot();
  });
});
