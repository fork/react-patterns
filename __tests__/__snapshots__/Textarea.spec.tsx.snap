// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Textarea should render correctly 1`] = `
<styled.div>
  <label
    htmlFor="1"
  >
    <styled.p
      as="span"
      variant="form-label"
    >
      Textarea Label
    </styled.p>
  </label>
  <textarea
    id="1"
    name="1"
  />
</styled.div>
`;

exports[`Textarea Textarea with default value should render correctly 1`] = `
<styled.div
  hideLabel={true}
>
  <label
    htmlFor="1"
  >
    <styled.p
      as="span"
      variant="form-label"
    >
      Textarea Label
    </styled.p>
  </label>
  <textarea
    id="1"
    name="1"
    value="true"
  />
</styled.div>
`;

exports[`Textarea Textarea with hidden label should render correctly 1`] = `
<styled.div
  hideLabel={true}
>
  <label
    htmlFor="1"
  >
    <styled.p
      as="span"
      variant="form-label"
    >
      Textarea Label
    </styled.p>
  </label>
  <textarea
    id="1"
    name="1"
  />
</styled.div>
`;
